tsdb/engine/wal/wal.go

```diff
@@ -1566,16 +1584,22 @@ type cacheEntry struct {
 //     []byte key
 //     []byte data
 //
-func marshalWALEntry(key []byte, timestamp int64, data []byte) []byte {
-	v := make([]byte, 8+4+4, 8+4+4+len(key)+len(data))
-	binary.BigEndian.PutUint64(v[0:8], uint64(timestamp))
-	binary.BigEndian.PutUint32(v[8:12], uint32(len(key)))
-	binary.BigEndian.PutUint32(v[12:16], uint32(len(data)))
-
-	v = append(v, key...)
-	v = append(v, data...)
+func marshalWALEntry(buf *bytes.Buffer, key []byte, timestamp int64, data []byte) {
+	// bytes.Buffer can't error, so ignore error checking in this code
+	var tmpbuf [8]byte
+	binary.BigEndian.PutUint64(tmpbuf[:], uint64(timestamp))
+	buf.Write(tmpbuf[:])
+	binary.BigEndian.PutUint32(tmpbuf[:4], uint32(len(key)))
+	buf.Write(tmpbuf[:4])
+	binary.BigEndian.PutUint32(tmpbuf[:4], uint32(len(data)))
+	buf.Write(tmpbuf[:4])
```

Comment:
The binary.Write uses reflection to figure out the data type. You have to make an allocation if you use it.